File: iOS/Sudoku/BoardView.swift
//
//  BoardView.swift
//  Sudoku
//

import SwiftUI

struct BoardView: View {
    @ObservedObject var game: SudokuGame
    /// When non-nil, the tutor is active: cells receive tutor tints and the
    /// usual selection / matching / highlighting tints are suppressed so the
    /// step's narration stays visually unambiguous.
    var tutorHighlights: [TutorHighlight]? = nil

    private var tutorActive: Bool { tutorHighlights != nil }

    /// Resolves the tutor decoration for a single cell. The cell's background
    /// tint comes from the highest-priority kind among matching highlights
    /// (target > eliminator > focus). The candidate-digit color map lets a
    /// single cell tint different digits differently — needed for hidden
    /// pair, where the pair digits stay (green) and the others get crossed
    /// out (red) all in the same cell.
    private func tutorInfo(row: Int, col: Int) -> (kind: TutorHighlight.Kind, colors: [Int: Color])? {
        guard let highlights = tutorHighlights else { return nil }
        var bestKind: TutorHighlight.Kind?
        var digitKinds: [Int: TutorHighlight.Kind] = [:]
        for h in highlights where h.row == row && h.col == col {
            switch h.kind {
            case .target:
                bestKind = .target
            case .eliminator:
                if bestKind != .target { bestKind = .eliminator }
            case .focus:
                if bestKind == nil { bestKind = .focus }
            }
            // Per-digit kind: target wins over eliminator for the same digit.
            for d in h.candidates {
                let existing = digitKinds[d]
                if existing == nil || (existing == .eliminator && h.kind == .target) {
                    digitKinds[d] = h.kind
                }
            }
        }
        guard let kind = bestKind else { return nil }
        let colors: [Int: Color] = digitKinds.compactMapValues { k in
            switch k {
            case .target: return .green
            case .eliminator: return .red
            case .focus: return nil  // focus doesn't tint candidates
            }
        }
        return (kind, colors)
    }

    var body: some View {
        GeometryReader { geo in
            let side = min(geo.size.width, geo.size.height)
            let cellSide = side / 9

            ZStack(alignment: .topLeading) {
                // Cells
                VStack(spacing: 0) {
                    ForEach(0..<9, id: \.self) { row in
                        HStack(spacing: 0) {
                            ForEach(0..<9, id: \.self) { col in
                                let info = tutorInfo(row: row, col: col)
                                CellView(
                                    cell: game.cells[row][col],
                                    isSelected: !tutorActive && game.selected?.row == row && game.selected?.col == col,
                                    isHighlighted: !tutorActive && (game.isHighlighted(row: row, col: col)
                                        || game.isUnavailableForSelectedValue(row: row, col: col)),
                                    isMatching: !tutorActive && game.isMatchingNumber(row: row, col: col),
                                    isError: !tutorActive && game.highlightMistakes && game.hasConflict(row: row, col: col),
                                    isLocked: game.isLocked(row: row, col: col),
                                    tutorHighlight: info?.kind,
                                    tutorCandidateColors: info?.colors ?? [:]
                                )
                                .frame(width: cellSide, height: cellSide)
                                .onTapGesture {
                                    if !tutorActive { game.select(row: row, col: col) }
                                }
                            }
                        }
                    }
                }

                // Grid lines
                gridLines(side: side)

                if game.isPaused {
                    pauseCover(side: side)
                }
            }
            .frame(width: side, height: side)
            .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .center)
        }
        .aspectRatio(1, contentMode: .fit)
    }

    private func pauseCover(side: CGFloat) -> some View {
        ZStack {
            Rectangle()
                .fill(Color(.systemBackground))
            VStack(spacing: 12) {
                Image(systemName: "pause.circle.fill")
                    .font(.system(size: 56))
                    .foregroundStyle(.secondary)
                Text("Paused")
                    .font(.title3.weight(.semibold))
                    .foregroundStyle(.secondary)
                Button("Resume") { game.togglePause() }
                    .buttonStyle(.borderedProminent)
            }
        }
        .frame(width: side, height: side)
        .contentShape(Rectangle())
    }

    private func gridLines(side: CGFloat) -> some View {
        let cell = side / 9
        return ZStack {
            // Thin lines for every cell boundary.
            ForEach(0...9, id: \.self) { i in
                let isThick = i % 3 == 0
                let width: CGFloat = isThick ? 2 : 0.5
                let color: Color = isThick ? .primary : .secondary

                // Horizontal
                Rectangle()
                    .fill(color)
                    .frame(width: side, height: width)
                    .position(x: side / 2, y: cell * CGFloat(i))

                // Vertical
                Rectangle()
                    .fill(color)
                    .frame(width: width, height: side)
                    .position(x: cell * CGFloat(i), y: side / 2)
            }
        }
        .allowsHitTesting(false)
    }
}

File: Android/app/src/main/java/com/derekgillett/sudoku/ui/BoardView.kt
package com.derekgillett.sudoku.ui

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.PlayArrow
import androidx.compose.material3.Button
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.derekgillett.sudoku.state.GameState
import com.derekgillett.sudoku.state.Highlights
import com.derekgillett.sudoku.state.TutorHighlight

@Composable
fun BoardView(
    state: GameState,
    onSelect: (Int, Int) -> Unit,
    onResume: () -> Unit,
    modifier: Modifier = Modifier,
    /** When non-null, the tutor is active. The board renders tutor tints
     * and suppresses the normal selection / matching / highlight colors so
     * the step's narration stays visually unambiguous. */
    tutorHighlights: List<TutorHighlight>? = null
) {
    val tutorActive = tutorHighlights != null
    BoxWithConstraints(
        modifier = modifier
            .fillMaxWidth()
            .aspectRatio(1f)
    ) {
        val boardModifier = Modifier
            .fillMaxSize()
            .background(MaterialTheme.colorScheme.surface)

        Column(modifier = boardModifier) {
            for (r in 0 until 9) {
                Row(
                    modifier = Modifier
                        .weight(1f)
                        .fillMaxWidth()
                ) {
                    for (c in 0 until 9) {
                        val cell = state.cells[r][c]
                        val info = tutorHighlights?.let { tutorInfo(it, r, c) }
                        val sel = state.selected
                        val isSelected = !tutorActive && sel?.row == r && sel.col == c
                        val isHighlighted = !tutorActive && (
                            Highlights.isHighlighted(state, r, c) ||
                                Highlights.isUnavailableForSelectedValue(state, r, c)
                            )
                        val isMatching = !tutorActive && Highlights.isMatchingNumber(state, r, c)
                        val isError = !tutorActive && state.highlightMistakes &&
                            Highlights.hasConflict(state, r, c)
                        val isLocked = Highlights.isLocked(state, r, c)

                        Box(
                            modifier = Modifier
                                .weight(1f)
                                .fillMaxHeight()
                                .clickable(enabled = !tutorActive) { onSelect(r, c) }
                        ) {
                            CellView(
                                cell = cell,
                                isSelected = isSelected,
                                isHighlighted = isHighlighted,
                                isMatching = isMatching,
                                isError = isError,
                                isLocked = isLocked,
                                tutorHighlight = info?.kind,
                                tutorCandidateColors = info?.colors ?: emptyMap()
                            )
                        }
                    }
                }
            }
        }

        // Grid lines overlay
        val lineColor = MaterialTheme.colorScheme.onSurface
        val secondaryColor = MaterialTheme.colorScheme.outline
        Canvas(modifier = Modifier.fillMaxSize()) {
            val cell = size.width / 9f
            for (i in 0..9) {
                val isThick = i % 3 == 0
                val width = if (isThick) 2f else 0.5f
                val color = if (isThick) lineColor else secondaryColor
                drawLine(
                    color = color,
                    start = androidx.compose.ui.geometry.Offset(0f, i * cell),
                    end = androidx.compose.ui.geometry.Offset(size.width, i * cell),
                    strokeWidth = width
                )
                drawLine(
                    color = color,
                    start = androidx.compose.ui.geometry.Offset(i * cell, 0f),
                    end = androidx.compose.ui.geometry.Offset(i * cell, size.height),
                    strokeWidth = width
                )
            }
        }

        if (state.isPaused) {
            PauseCover(onResume = onResume)
        }
    }
}

private data class TutorCellInfo(val kind: TutorHighlight.Kind, val colors: Map<Int, Color>)

/**
 * Resolves the tutor decoration for a single cell. Background tint comes
 * from the highest-priority highlight kind (target > eliminator > focus).
 * The candidate-digit color map lets one cell tint different digits
 * differently — needed for hidden pair, where pair digits stay (green) and
 * the others get crossed out (red) all in the same cell.
 */
private fun tutorInfo(highlights: List<TutorHighlight>, row: Int, col: Int): TutorCellInfo? {
    var bestKind: TutorHighlight.Kind? = null
    val digitKinds = HashMap<Int, TutorHighlight.Kind>()
    for (h in highlights) {
        if (h.row != row || h.col != col) continue
        bestKind = when (h.kind) {
            TutorHighlight.Kind.TARGET -> TutorHighlight.Kind.TARGET
            TutorHighlight.Kind.ELIMINATOR ->
                if (bestKind != TutorHighlight.Kind.TARGET) TutorHighlight.Kind.ELIMINATOR else bestKind
            TutorHighlight.Kind.FOCUS ->
                bestKind ?: TutorHighlight.Kind.FOCUS
        }
        for (d in h.candidates) {
            val existing = digitKinds[d]
            // Target wins over eliminator for the same digit.
            if (existing == null || (existing == TutorHighlight.Kind.ELIMINATOR && h.kind == TutorHighlight.Kind.TARGET)) {
                digitKinds[d] = h.kind
            }
        }
    }
    val kind = bestKind ?: return null
    val colors = digitKinds.mapNotNull { (d, k) ->
        when (k) {
            TutorHighlight.Kind.TARGET -> d to Color(0xFF2E7D32)
            TutorHighlight.Kind.ELIMINATOR -> d to Color(0xFFD32F2F)
            TutorHighlight.Kind.FOCUS -> null
        }
    }.toMap()
    return TutorCellInfo(kind, colors)
}

@Composable
private fun PauseCover(onResume: () -> Unit) {
    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(MaterialTheme.colorScheme.surface)
            .clickable(enabled = true, onClick = {}),
        contentAlignment = Alignment.Center
    ) {
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(12.dp)
        ) {
            Icon(
                imageVector = Icons.Filled.PlayArrow,
                contentDescription = null,
                modifier = Modifier.size(56.dp),
                tint = MaterialTheme.colorScheme.onSurfaceVariant
            )
            Text(
                "Paused",
                fontSize = 20.sp,
                fontWeight = FontWeight.SemiBold,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
            Button(onClick = onResume) {
                Text("Resume")
            }
        }
    }
}
